packages/tx-construction/src/createTransactionInternals.ts

```diff
@@ -14,34 +15,56 @@ export type CreateTxInternalsProps = {
   requiredExtraSignatures?: Crypto.Ed25519KeyHashHex[];
 };
 
-export const createTransactionInternals = async ({
+/** Returns transaction body based on the information available before input selection takes place */
+export const createPreInputSelectionTxBody = ({
   auxiliaryData,
   withdrawals,
   certificates,
   validityInterval,
-  inputSelection,
   collaterals,
   mint,
   scriptIntegrityHash,
-  requiredExtraSignatures
-}: CreateTxInternalsProps): Promise<Cardano.TxBodyWithHash> => {
+  requiredExtraSignatures,
+  outputs
+}: Omit<CreateTxInternalsProps, 'inputSelection'> & { outputs?: Cardano.TxOut[] }): {
+  body: TxBodyPreInputSelection;
+  auxiliaryData?: Cardano.AuxiliaryData;
+} => {
+  const body: TxBodyPreInputSelection = {
+    auxiliaryDataHash: auxiliaryData ? Cardano.computeAuxiliaryDataHash(auxiliaryData) : undefined,
+    certificates,
+    mint,
+    outputs: outputs || [],
+    requiredExtraSignatures,
+    scriptIntegrityHash,
+    validityInterval,
+    ...(withdrawals?.length && { withdrawals }),
+    ...(collaterals && { collaterals: [...collaterals] })
+  };
+
+  return {
+    auxiliaryData,
+    body
+  };
+};
+
+/** Updates the txBody after input selection takes place with the calculated change and selected inputs */
+export const includeChangeAndInputs = async ({
+  bodyPreInputSelection,
+  inputSelection
+}: Pick<CreateTxInternalsProps, 'inputSelection'> & {
+  bodyPreInputSelection: TxBodyPreInputSelection;
+}): Promise<Cardano.TxBodyWithHash> => {
   const outputs = [...inputSelection.outputs];
   for (const changeOutput of inputSelection.change) {
     outputs.push(changeOutput);
   }
```

Comment:
The diff clearly says these lines wasn't touched in this PR, but since we are refactoring this piece of code why not ...
```suggestion
  const outputs = [...inputSelection.outputs, ...inputSelection.change];
```
... or, even better, inline in `body` definition in next line?

Reply:
Done here: [ec76f62](https://github.com/input-output-hk/cardano-js-sdk/pull/1099/commits/ec76f62e8dff3be6334bd8bac694f380c9061544)